include/time_series/interface.hpp

```diff
@@ -0,0 +1,107 @@
+/**
+ * @file interface.hpp
+ * @author Vincent Berenz
+ * license License BSD-3-Clause
+ * @copyright Copyright (c) 2019, Max Planck Gesellshaft.
+ */
+
+#pragma once
+
+#include <cstddef>
+#include <limits>
+
+namespace time_series
+{
+typedef long int Index;
+typedef long double Timestamp;
+
+/**
+ * \brief Interface for time series.
+ * A time_series implements  \f$ X_{{oldest}:{newest}} \f$ which can
+ * safely be accessed from either multiple threads or multiple processes.
+ *
+ * this object has the following properties:
+ * - an oldest timeindex \f$ oldest\f$,
+ * - a newest timeindex \f$ newest \f$,
+ * - a value \f$ X_i \f$ for each  \f$ i \in \{oldest, oldest + 1 , ...,
+ * newest\} \f$,
+ * - a length \f$length\f$
+ * - and a maximum length \f$maxlength\f$
+ */
+template <typename T>
+class TimeSeriesInterface
+{
+public:
+    /*! \brief returns \f$ newest \f$. waits if the time_series is empty.
+     */
+    virtual Index newest_timeindex() = 0;
+
+    /*! \brief returns the number of element contained in the queue.
+     */
+    virtual Index count_appended_elements() = 0;
+
+    /*! \brief returns \f$ oldest \f$. waits if the time_series is empty.
+     */
+    virtual Index oldest_timeindex() = 0;
+
+    /*! \brief returns \f$ X_{newest} \f$. waits if the time_series is empty.
+     */
+    virtual T newest_element() = 0;
+
+    /*! \brief returns \f$ X_{timeindex} \f$. waits if the time_series is empty
+     * or if \f$timeindex > newest \f$.
+     */
+    virtual T operator[](const Index &timeindex) = 0;
+
+    /*! \brief returns the time in miliseconds when \f$ X_{timeindex} \f$
+     * was appended. Waits if the time_series is empty
+     * or if \f$timeindex > newest \f$.
+     */
+    virtual Timestamp timestamp_ms(const Index &timeindex) = 0;
+
+    /*! \brief returns the time in seconds when \f$ X_{timeindex} \f$
+     * was appended. Waits if the time_series is empty
+     * or if \f$timeindex > newest \f$.
+     */
+    virtual Timestamp timestamp_s(const Index &timeindex) = 0;
+
+    /*! \brief Wait until the defined time index is reached. If the input time
+     *  is below the oldest time index that have been registered read an
+     * exception
+     * is return.
```

Comment:
I think this sentence is somehow broken.

Reply:
+1